src/Altinn.App.Core/Internal/Data/CachedInstanceDataAccessor.cs

```diff
@@ -40,120 +49,314 @@ IAppModel appModel
         }
-        if (dataTypeObj.MaxCount != 1)
+
+        return dataType;
+    }
+
+    /// <inheritdoc />
+    public void AddFormDataElement(string dataTypeString, object data)
+    {
+        var dataType = GetDataTypeByString(dataTypeString).Result;
+        if (dataType.AppLogic?.ClassRef is not { } classRef)
         {
-            throw new InvalidOperationException($"Data type {dataType} is not a single data type");
+            throw new InvalidOperationException(
+                $"Data type {dataTypeString} does not have a class reference in app metadata"
+            );
         }
-        var dataElement = Instance.Data.Find(d => d.DataType == dataType);
-        if (dataElement == null)
+
+        var modelType = data.GetType();
+        if (modelType.FullName != classRef)
         {
-            return null;
+            throw new InvalidOperationException(
+                $"Data object registered for {dataTypeString} is not of type {classRef} as specified in application metadata"
+            );
         }
 
-        return await GetData(dataElement);
+        var (bytes, contentType) = _modelSerializationService.SerializeToStorage(data, dataType);
+
+        _dataElementsToAdd.Add((dataType, contentType, null, bytes));
+        // var dataElement = await _dataClient.InsertBinaryData(
+        //     Instance.Id,
+        //     dataTypeString,
+        //     contentType,
+        //     null,
+        //     new MemoryAsStream(binaryData)
+        // );
+        // Instance.Data.Add(dataElement);
+        //
+        // return dataElement;
     }
 
     /// <inheritdoc />
-    public async Task<object> GetData(DataElementId dataElementId)
+    public void AddAttachmentDataElement(
+        string dataTypeString,
+        string contentType,
+        string? filename,
+        ReadOnlyMemory<byte> bytes
+    )
     {
-        return await _cache.GetOrCreate(
-            dataElementId,
-            async _ =>
+        var dataType = GetDataTypeByString(dataTypeString).Result;
+        if (dataType.AppLogic?.ClassRef is not null)
+        {
+            throw new InvalidOperationException(
+                $"Data type {dataTypeString} has a AppLogic.ClassRef in app metadata, and is not a binary data element"
+            );
+        }
+        _dataElementsToAdd.Add((dataType, contentType, filename, bytes));
+    }
+
+    /// <inheritdoc />
+    public void RemoveDataElement(DataElementId dataElementId)
+    {
+        var idAsString = dataElementId.ToString();
+        var dataElement = Instance.Data.Find(d => d.Id == idAsString);
+        if (dataElement is null)
+        {
+            throw new InvalidOperationException($"Data element with id {idAsString} not found in instance");
+        }
+        //TODO: Add to list of data elements to delete
+        // await _dataClient.DeleteData(_org, _app, _instanceOwnerPartyId, _instanceGuid, dataElementId.Guid, true);
+
+        Instance.Data.Remove(dataElement);
```

Comment:
Is this meant to use `_dataElementsToDelete` instead?

Reply:
Thanks!